crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs

```diff
@@ -17,84 +18,253 @@ pub(crate) enum PeerDiscoverResult {
+            if near_nodes.is_empty() {
+                // maybe just wait and try again?
+                backoff.wait().await;
+                continue;
             }
 
-            interval_ms *= 2;
-            if interval_ms > timeout_ms - elapsed_ms {
-                interval_ms = timeout_ms - elapsed_ms;
+            // the next step involves making network requests
+            // so check our timeout first
+            timeout.ok()?;
+
+            // shuffle the returned nodes so we don't keep hammering the same one
+            use rand::prelude::*;
+            near_nodes.shuffle(&mut rand::thread_rng());
+
+            let mut added_data = false;
+            for node in near_nodes {
+                // try connecting to the returned nodes
+                if let PeerDiscoverResult::OkRemote { con_hnd, .. } =
+                    peer_connect(inner.clone(), &node, timeout).await
+                {
+                    // make a peer query for the basis
+                    let payload = wire::Wire::peer_query(inner.space.clone(), basis_loc);
+                    match con_hnd.request(&payload, timeout).await {
+                        Ok(wire::Wire::PeerQueryResp(wire::PeerQueryResp { peer_list })) => {
+                            if peer_list.is_empty() {
+                                tracing::warn!("empty discovery peer list");
+                                continue;
+                            }
+                            // if we got results, add them to our peer store
+                            for agent_info_signed in peer_list {
+                                let agent = agent_info_signed.agent.clone();
+                                if let Err(err) = inner
+                                    .evt_sender
+                                    .put_agent_info_signed(PutAgentInfoSignedEvt {
+                                        space: inner.space.clone(),
+                                        agent,
+                                        agent_info_signed,
+                                    })
+                                    .await
+                                {
+                                    tracing::error!(?err, "error storing peer_queried agent_info");
+                                }
+                            }
+                            // then break, to pull up the local query
+                            // that should now include these new results
+                            added_data = true;
+                            break;
+                        }
+                        peer_resp => {
+                            tracing::warn!(?peer_resp, "unexpected peer resp");
+                        }
+                    }
+                }
             }
 
-            tokio::time::sleep(std::time::Duration::from_millis(interval_ms)).await;
+            if !added_data {
+                backoff.wait().await;
```

Comment:
So this is happening in the context of a single_rpc. Is an rpc able to wait long enough to handle exponentially backing off?

Reply:
The backoff itself doesn't have a problem... the single_rpc timeout will be something like 30 seconds, and this backoff starts at 100ms up to a max of 1000ms... so it should get quite a few tries in...

EXCEPT... each internal call to things like `peer_connect` will use the full remaining timeout, so if the first returned node data we get happens to be to a node that is offline, we'll end up timing out without trying any additional nodes.

Again, I'm not sure a better way to solve this without making requests just dangle indefinitely if we've got that QUIC idle timeout set to 30 seconds...

Reply:
We're going to call this okay for now, we may need to address this differently in the future.